### PR TITLE
utils test coverage improvements

### PR DESCRIPTION
**Pull request by sr-murthy**
Addresses #2273 

## Pull Request Checklist

- [x] A news fragment is added in `news/` describing what is new.
- [Y ] Test cases added for changed code.

## Describe what you have changed in this PR.

* Update `pyproject.toml` to add a test coverage entrypoint
* Add tests for `pdm.utils`

### CODE REVIEW

**Comment by sr-murthy** [comment]
@frostming This PR is complete, including the news fragment. It adds 7% to test coverage in `utils` (latest measurement is 87%), and adds 1% to global test coverage. This is branch-based measurement, so untested branches in existing tests are considered missing. There were some untested functions in `utils` which I didn't touch, because I wasn't sure about, and I've pre-existing tests alone.

So existing test coverage on `main` for `utils` and the global coverage is:

    ---------- coverage: platform darwin, python 3.11.2-final-0 ----------
    Name                                               Stmts   Miss Branch BrPart  Cover   Missing
    ----------------------------------------------------------------------------------------------
    ...
    src/pdm/utils.py                                     256     44     88     16    80%   37-38, 48, 90, 106, 117, 120-121, 124-125,
    132->139, 135->139, 151-153, 191-195, 210, 262-263, 286, 292, 296, 299, 314-316, 321->320, 373-374, 403, 405-412, 431-436, 456-457
    ----------------------------------------------------------------------------------------------
    TOTAL                                              10200   1427   4567    505    83%

With this PR the measurement is:

    ---------- coverage: platform darwin, python 3.11.2-final-0 ----------
    Name                                               Stmts   Miss Branch BrPart  Cover   Missing
    ----------------------------------------------------------------------------------------------
    ...
    src/pdm/utils.py                                     256     29     86      8    87%   40-41, 51, 90, 106, 151-153, 193, 210,
    314-316, 321->320, 403, 405-412, 431-436, 456-457
    ----------------------------------------------------------------------------------------------
    TOTAL                                              10200   1412   4565    497    84%

**Review by frostming — APPROVED** [review]
Cool, LGTM 🎉

File: src/pdm/utils.py [hunk 1]
```diff
@@ -69,10 +72,7 @@ def url_without_fragments(url: str) -> str:
 
 
 def join_list_with(items: list[Any], sep: Any) -> list[Any]:
-    new_items = []
-    for item in items:
-        new_items.extend([item, sep])
-    return new_items[:-1]
+    return list(iter_chain.from_iterable(iter_product(items, [sep])))[:-1]
```
Comment (frostming) [inline src/pdm/utils.py:75]: This brings more burden to understand, I prefer to keep the old code.

**Review by frostming — CHANGES_REQUESTED** [review]
Please fix the test failures on older Python versions and Windows.

File: src/pdm/utils.py (hunk 1 above)
Comment (sr-murthy) [reply src/pdm/utils.py:75]: OK, no problem. Will revert

Comment (sr-murthy) [reply src/pdm/utils.py]: These test failures relate to path processing, which obviously works differently on Windows and POSIX. I'll need to revisit some of the test data here.

**Comment by sr-murthy** [comment]
Will be pushing an update in a few days, hopefully will fix all of the Windows failures.

File: src/pdm/utils.py (hunk 1 above)
Comment (sr-murthy) [reply src/pdm/utils.py]: @frostming I think I've fixed the test failures. There were issues here: (1) incompatibility in test path instantiation between POSIX and Windows platforms which affected `utils.get_venv_like_prefix` and `utils.path_to_url`, (2) incompatibility in path equality checking between pre-Python 3.12.x and Python 3.12.x versions which affected `utils.get_venv_like_prefix`.

I have Python 3.12.a01 installed via `pyenv`, but was unable to install PDM(2.9.3) due to an error, so could not test the changes locally in 3.12. But they should now work.

Comment (frostming) [reply src/pdm/utils.py]: path comparison is still failing. Thanks

**Comment by sr-murthy** [comment]
OK, I might just skip tests for `path_to_url` on Windows.

**Comment by sr-murthy** [comment]
Tests updated to skip some for Windows.

File: tests/test_utils.py [hunk 2]
```diff
@@ -11,6 +14,192 @@
...
+            utils.url_to_path("not_a_file_scheme://netloc/path")
+
+    def test_non_windows_non_local_file_url(self):
+        with mock.patch("pdm.utils.sys.platform", "non_windows"):
+            with pytest.raises(ValueError):
+                utils.url_to_path("file://non_local_netloc/file/url")
+
+    def test_non_windows_localhost_local_file_url(self):
+        with mock.patch("pdm.utils.sys.platform", "non_windows"):
+            assert utils.url_to_path("file://localhost/local/file/path") == "/local/file/path"
```
Comment (frostming) [inline tests/test_utils.py:179]: Maybe skip it on windows and add another one for Windows only.

Comment (sr-murthy) [reply tests/test_utils.py:179]: Np.

Comment (sr-murthy) [reply tests/test_utils.py:181]: OK, I've added a single Windows-only test case for `utils.url_to_path` testing the conversion of a localhost local file URL to a Windows path. I hope this works - I don't use Windows and don't have access to a Windows Python environment, so it is tricky for me to test this.

**Comment by sr-murthy** [comment]
@frostming Assuming this now passes, should I rebase some commits to squash/tidy up?

File: tests/test_utils.py [hunk 3]
```diff
@@ -11,6 +14,199 @@
...
+        with pytest.raises(AssertionError):
+            utils.url_to_path("not_a_file_scheme://netloc/path")
+
+    @pytest.mark.skipif(sys.platform.startswith("win"), reason="Non-Windows test")
+    def test_non_windows_non_local_file_url(self):
+        with mock.patch("pdm.utils.sys.platform", "non_windows"):
+            with pytest.raises(ValueError):
+                utils.url_to_path("file://non_local_netloc/file/url")
+
+    @pytest.mark.skipif(sys.platform.startswith("win"), reason="Windows-only test")
```
Comment (frostming) [inline tests/test_utils.py:178]:
```suggestion
    @pytest.mark.skipif(sys.platform.startswith("win"), reason="Non-Windows test")
```

File: tests/test_utils.py [hunk 4]
```diff
@@ -11,6 +14,199 @@
...
+    def test_non_windows_localhost_local_file_url(self):
+        with mock.patch("pdm.utils.sys.platform", "non_windows"):
+            assert utils.url_to_path("file://localhost/local/file/path") == r"\local\file\path"
```
Comment (frostming) [inline tests/test_utils.py:181]:
```suggestion
            assert utils.url_to_path("file://localhost/local/file/path") == "/local/file/path"
```

**Comment by frostming** [comment]
> @frostming Assuming this now passes, should I rebase some commits to squash/tidy up?

We will squash when merge

**Pull request merged**